File: standard-udfs-avro/src/test/java/com/linkedin/stdudfs/avro/examples/TestStructCreateByIndexFunctionWrapper.java
package com.linkedin.stdudfs.avro.examples;

import com.google.common.collect.ImmutableList;
import org.apache.avro.Schema;
import org.testng.annotations.Test;

import static com.linkedin.stdudfs.avro.common.AssertAvroUdf.*;
import static com.linkedin.stdudfs.avro.common.SchemaFactory.*;


public class TestStructCreateByIndexFunctionWrapper {
  @Test
  public void testStructCreateByIndex() {
    assertFunction(new StructCreateByIndexFunctionWrapper(), new Schema[]{STRING, STRING}, new Object[]{"x", "y"},
        ImmutableList.of("x", "y"));

    assertFunction(new StructCreateByIndexFunctionWrapper(), new Schema[]{STRING, STRING}, new Object[]{null, "y"},
        null);

    assertFunction(new StructCreateByIndexFunctionWrapper(), new Schema[]{STRING, STRING}, new Object[]{"x", null},
        null);
  }
}

File: standard-udfs-examples/standard-udfs-example-udfs-spark/src/test/scala/com/linkedin/stdudfs/examples/spark/TestStructCreateByIndexFunctionWrapper.scala
/**
 * BSD 2-CLAUSE LICENSE
 *
 * Copyright 2018 LinkedIn Corporation.
 * All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package com.linkedin.stdudfs.examples.spark

import com.linkedin.stdudfs.spark.common.AssertSparkExpression._
import org.apache.spark.sql.Row
import org.testng.annotations.{BeforeClass, Test}

import scala.collection.mutable.WrappedArray._

class TestStructCreateByIndexFunctionWrapper {

  @BeforeClass
  def registerFunction(): Unit = {
    registerStandardUdf(
      "struct_create_by_index",
      classOf[StructCreateByIndexFunctionWrapper]
    )
  }

  @Test
  def testStructCreateByIndexFunction(): Unit = {
    assertFunction("struct_create_by_index('x', 'y')", Row("x", "y"))
    assertFunction("struct_create_by_index(1, array(1))", Row(1, make(Array(1))))
    assertFunction("struct_create_by_index(null, null)", null)
  }
}
